### PR TITLE
Execution Plan Controller Tests

### PR DESCRIPTION
**Pull request by laurennat**
Execution Plan controller tests
image

### CODE REVIEW

File: test/unit/executionPlanWebviewController.test.ts [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+            createExecutionPlanGraphsStub.firstCall.args,
+            [
+                mockState,
+                controller.executionPlanService,
+                [controller.executionPlanContents],
+            ],
+            "createExecutionPlanGraphs should be called with correct arguments",
+        );
+
+        createExecutionPlanGraphsStub.restore();
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:115]: For all these reducer tests, we also need to check to make sure that the state _after_ the reducer runs is correct.  Unless that's happening elsewhere?

File: test/unit/executionPlanWebviewController.test.ts [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+    });
+
+    test("should call saveExecutionPlan in saveExecutionPlan reducer", async () => {
+        const saveExecutionPlanStub = sandbox
+            .stub(epUtils, "saveExecutionPlan")
+            .resolves({
+                executionPlanState: {
+                    executionPlanGraphs: [],
+                    loadState: ApiStatus.Loaded,
+                    totalCost: 100,
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:125]: If there's some state that's shared across tests, you can pull it out into a shared const.  If the shape of the state changes in the future, it helps to minimize the number of spots that need updating

File: test/unit/executionPlanWebviewController.test.ts [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+
+    test("saveExecutionPlan: should call saveExecutionPlan and return the state", async () => {
+        const showSaveDialogStub = sinon.stub(vscode.window, "showSaveDialog");
+
+        const mockState = {
+            executionPlanState: {
+                loadState: ApiStatus.Loading,
+                executionPlanGraphs: [],
+                totalCost: 0,
+            },
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:325]: common mock starting state can be centralized

File: test/unit/executionPlanWebviewController.test.ts [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+        };
+
+        const mockPayload = { sqlPlanContent: executionPlanContents };
+
+        const mockUri = vscode.Uri.file("/mock/path/to/plan.sqlplan");
+
+        showSaveDialogStub.resolves(mockUri);
+
+        const result = await epUtils.saveExecutionPlan(mockState, mockPayload);
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:334]: this should also check that the file has been written to disk, no?

File: test/unit/executionPlanWebviewController.test.ts [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+
+        assert.strictEqual(
+            result,
+            0,
+            "Total cost should be 0 for an empty executionPlanGraphs array",
+        );
+    });
+
+    test("formatXml: should return original xml contents if it is not a valid xml file", () => {
+        const invalidXml = "</";
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:550]: What happens when I attempt to open a sqlplan file with invalid XML?  Does it display an error message to the user?

File: test/unit/executionPlanWebviewController.test.ts [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+            extensionPath: "path",
+        } as unknown as vscode.ExtensionContext;
+
+        mockExecutionPlanService =
+            sandbox.createStubInstance(ExecutionPlanService);
+        mockUntitledSqlDocumentService = sandbox.createStubInstance(
+            UntitledSqlDocumentService,
+        );
+
+        mockState = {
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:45]: better name like `mockInitialState` would help readability

File: test/unit/executionPlanWebviewController.test.ts [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+            createExecutionPlanGraphsStub.firstCall.args,
+            [
+                mockState,
+                controller.executionPlanService,
+                [controller.executionPlanContents],
+            ],
+            "createExecutionPlanGraphs should be called with correct arguments",
+        );
+
+        assert.deepStrictEqual(result, mockResultState);
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:113]: add message for failures (and other spots)

File: test/unit/executionPlanWebviewController.test.ts [hunk 8]
```diff
@@ -229,11 +249,15 @@ suite("ExecutionPlanWebviewController", () => {
...
-            [mockState, mockPayload],
+            [mockInitialState, mockPayload],
             "showQuery should be called with correct arguments",
         );
 
-        assert.deepStrictEqual(result, mockResultState);
+        assert.deepStrictEqual(
+            result,
+            mockResultState,
+            "State should be updated by reducer call",
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:259]: Even better would be to say _how_ the state should've been updated (e.g. "state should have totalCost updated")

File: test/unit/executionPlanWebviewController.test.ts [hunk 9]
```diff
@@ -420,29 +429,30 @@ suite("Execution Plan Utilities", () => {
...
+        );
         assert.deepStrictEqual(
             result.executionPlanState.loadState,
             ApiStatus.Error,
+            "The load state should be updated",
         );
         assert.deepStrictEqual(
             result.executionPlanState.errorMessage,
             "Mock Error",
+            "The correct error message should be updated in state",
```
Comment (Benjin) [inline test/unit/executionPlanWebviewController.test.ts:445]: Perfect!

**Review by Benjin — APPROVED** [review]

**Review by lewis-sanchez — APPROVED** [review]

**Review by Benjin — APPROVED** [review]

**Pull request merged**